### PR TITLE
Deprecate None values for Date, Datetime and Time trait types

### LINKED ISSUE #1438: Change default for allow_none and allow_datetime on Date and related Traits
The `Date` trait currently accepts both `None` and `datetime` objects unless explicitly asked not to using `allow_none=False` or `allow_datetime=False` (respectively).

Similarly, the `Time` and `Datetime` traits both currently accept `None` unless `allow_none=False` is explicitly given. In effect, `allow_datetime` (for `Date`) and `allow_none` (for all three datetime-related trait types) default to `True`, for backwards compatibility.

We propose to make the default for both `allow_none` and `allow_datetime` in these contexts `False` instead of `True`. This would be a backwards incompatible change, so would require deprecation warnings. Specifically, the proposal would be:

- In Traits 6.3, issue a `DeprecationWarning` if the `allow_none` metadata is not given and a `None` value is assigned to a `Date`, `Datetime` or `Time` trait. That warning would explain that if you want `None` to be explicitly permitted, you should use `allow_none=True`
- Similarly in Traits 6.3, issue a `DeprecationWarning` if an instance of `datetime` is assigned to a `Date` trait and `allow_datetime` was not specified in the trait definition. Again, the warning would explain that you need to use `allow_datetime=True` to explicitly allow datetime instances (or use an `Instance` or a `Union` trait).
- In some future version of Traits, perhaps Traits 7.0 depending on timing, change the defaults for `allow_none` and `allow_datetime` and remove the `DeprecationWarning`s.

### PR DESCRIPTION
**Pull request by mdickinson**
This PR builds on #1441 to also deprecate acceptance of `None` values for `Date` traits. While I think this is the right way to go, ~this does introduce a discrepancy with the `Datetime` and `Time` classes, where `None` is still accepted by default with no deprecation warning issued.~

Update: the PR has been extended to cover `Datetime` and `Time` classes; this required realising the `Datetime` and `Time` trait types (so that they're no longer instances of `BaseInstance`).

Closes #1438

**Checklist**
- [x] Tests
- [x] Update API reference (`docs/source/traits_api_reference`)  (relevant docstrings updated)
- [ ] ~Update User manual (`docs/source/traits_user_manual`)~ N/A
- [x] Update type annotation hints in `traits-stubs`

### CODE REVIEW

File: traits/tests/test_time.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ class HasTimeTraits(HasStrictTraits):
     none_allowed = Time(allow_none=True)
 
 
-class TestDatetime(unittest.TestCase):
+class TestTime(unittest.TestCase):
```
Comment (mdickinson) [inline traits/tests/test_time.py:47]: Drive-by copypasta fix, unrelated to the main purpose of the PR

File: traits/trait_types.py [hunk 2]
```diff
@@ -4321,8 +4339,7 @@ def validate(self, object, name, value):
                         "In the future, datetime.datetime instances will no "
                         "longer be accepted by this trait type. To accept "
                         "datetimes and silence this warning, use "
-                        "Date(allow_datetime=True) or "
-                        "Union(Datetime(), Date())."
+                        "Date(allow_datetime=True) or Union(Datetime, Date)."
```
Comment (rahulporuri) [inline traits/trait_types.py:4342]: i'm not sure why this change was made.

**Review by rahulporuri — APPROVED** [review]
LGTM with one minor question. We already have the issue regd documenting `Date`, `Datetime` and `Time` so there's no documentation changes that need to be made in this PR.

Just to confirm, can there be any issues rewriting the traits as subclasses of `TraitType` instead of using `BaseInstance`?

Finally, how do we plan to track the work needed for traits 7.0.0? Do we want to create an issue now/a milestone?

File: traits/trait_types.py (hunk 2 above)
Comment (mdickinson) [reply traits/trait_types.py:4342]: It shouldn't have been; I resolved a merge conflict with master the wrong way around.

**Comment by mdickinson** [comment]
> Just to confirm, can there be any issues rewriting the traits as subclasses of `TraitType` instead of using `BaseInstance`?

Yes. There are potential behaviour changes here, but mostly in a good direction: that is, treating an instance of `TraitType` as though it were a subclass of `TraitType` is fraught with issues.

For example, if anyone is "instantiating" `Datetime` or `Time` in their code and making use of the particular keyword arguments that `BaseInstance` understands, their code will change behaviour.

But I checked existing code (to the extent that that code was available to me) for usage patterns, and I would expect the vast majority of existing code to be unaffected by this change (beyond the intentional part of the change - the new `DeprecationWarning`).

**Comment by mdickinson** [comment]
> Finally, how do we plan to track the work needed for traits 7.0.0? Do we want to create an issue now/a milestone?

The milestone already exists, and has done for some time. I already have an issue in that milestone as a reminder to review deprecations.

File: traits/trait_types.py (hunk 2 above)
Comment (mdickinson) [reply traits/trait_types.py]: Fixed.

**Comment by mdickinson** [comment]
> I already have an issue in that milestone as a reminder to review deprecations.

#1268
